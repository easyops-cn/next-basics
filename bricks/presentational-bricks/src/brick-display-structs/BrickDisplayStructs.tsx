import React from "react";
import { get, isObject } from "lodash";
import { useTranslation } from "react-i18next";
import { NS_PRESENTATIONAL_BRICKS, K } from "../i18n/constants";
import { FieldToDisplay } from "./index";

interface BrickDisplayStructsProps {
  // object or array
  value: any;
  displayType: "stringify" | FieldToDisplay;
  emptyText: string;
}

function stringify(value: any): string {
  return JSON.stringify(value);
}

function partField(
  value: any,
  displayType: FieldToDisplay,
  emptyText: string,
  error: string
): string {
  let result = "";

  if (value === undefined) {
    return emptyText || error;
  } else if (Array.isArray(value)) {
    result = value
      .map((item) => {
        return displayType.field
          ? get(item, displayType.field, error)
          : isObject(item)
          ? stringify(item)
          : item;
      })
      .join(displayType.separator);
  } else {
    result = get(value, displayType.field, error);
  }

  return result;
}

export function BrickDisplayStructs(
  props: BrickDisplayStructsProps
): React.ReactElement {
  const { t } = useTranslation(NS_PRESENTATIONAL_BRICKS);
  const error = t(K.UNKNOWN_ERROR);

  const text =
    props.displayType === "stringify"
      ? stringify(props.value)
      : partField(props.value, props.displayType, props.emptyText, error);

  return <span>{text}</span>;
}
