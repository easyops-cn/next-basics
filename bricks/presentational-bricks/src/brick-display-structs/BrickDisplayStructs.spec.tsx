import React from "react";
import { shallow } from "enzyme";
import { BrickDisplayStructs } from "./BrickDisplayStructs";
import { FieldToDisplay } from "./index";

describe("BrickDisplayStructs", () => {
  it("should work, stringify", () => {
    const value = { text: "hello, world" };
    const displayType = "stringify";
    const wrapper = shallow(
      <BrickDisplayStructs value={value} displayType={displayType} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should work, field to display", () => {
    let value: any = [
      { name: "irelia", email: "irelia@ionia.island" },
      { name: "akali", email: "akali@ionia.island" },
    ];
    const displayType: FieldToDisplay = {
      field: "name",
      separator: "; ",
    };
    let wrapper = shallow(
      <BrickDisplayStructs value={value} displayType={displayType} />
    );
    const span = wrapper.find("span");
    expect(span.childAt(0).contains("irelia; akali")).toBe(true);
    wrapper.setProps({
      displayType: { separator: "; " },
      value: [1, 2, { name: "akali" }],
    });
    wrapper.update();
    expect(wrapper.childAt(0).text()).toBe('1; 2; {"name":"akali"}');

    value = { text: "hello, world" };
    wrapper = shallow(
      <BrickDisplayStructs value={value} displayType={displayType} />
    );
    expect(wrapper.contains("UNKNOWN_ERROR")).toBe(true);
  });
});
