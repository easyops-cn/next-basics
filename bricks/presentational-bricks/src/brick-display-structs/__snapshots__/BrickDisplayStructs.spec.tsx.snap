// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrickDisplayStructs should work, stringify 1`] = `
<span>
  {"text":"hello, world"}
</span>
`;
